ext/page_log/palite/palite.cpp

```diff
@@ -1019,741 +799,1190 @@ class Storage {
+        return acc.table.get_table_ids(acc.conn);
+    }
 
-        return 0;
+    void
+    put_checkpoint(uint64_t lsn, uint64_t checkpoint_timestamp, const WT_ITEM *checkpoint_metadata)
+    {
+        auto acc = request<Checkpoints, AccessMode::WRITE>();
+        acc.table.put(acc.conn, lsn, checkpoint_timestamp, checkpoint_metadata);
     }
 
     int
-    get_page_ids(Connection &conn, uint64_t checkpoint_lsn, uint64_t table_id, WT_ITEM *page_ids,
-      size_t *page_count)
+    get_checkpoint(uint64_t &lsn, uint64_t *timestamp, WT_ITEM *checkpoint_metadata)
     {
-        StatementPtr stmt = conn.db_statement(Statement::GET_PAGE_IDS);
-        SQ_CHECK(sqlite3_bind_int64, stmt.get(), 1, static_cast<sqlite3_int64>(table_id));
-        SQ_CHECK(sqlite3_bind_int64, stmt.get(), 2, static_cast<sqlite3_int64>(checkpoint_lsn));
+        auto acc = request<Checkpoints, AccessMode::READ>();
 
-        std::vector<uint64_t> ids;
-        while (SQ_CHECK(sqlite3_step, stmt.get()) == SQLITE_ROW) {
-            ids.push_back(sqlite3_column_int64(stmt.get(), 0));
-        }
-        LOG_DEBUG("Found {} page IDs for table_id={} at checkpoint_lsn={}", ids.size(), table_id,
-          checkpoint_lsn);
-        LOG_DIAG("Page IDs: {}", join(ids, ", "));
+        return acc.table.get(acc.conn, lsn, timestamp, checkpoint_metadata);
+    }
 
-        *page_count = ids.size();
-        if (ids.size() > 0) {
-            fill_item(page_ids, ids.data(), ids.size() * sizeof(uint64_t));
-        }
+    void
+    delete_checkpoints(uint64_t lsn)
+    {
+        auto acc = request<Checkpoints, AccessMode::WRITE>();
+        acc.table.delete_many(acc.conn, lsn);
+    }
 
-        return 0;
+    void
+    put_page(uint64_t table_id, uint64_t page_id, uint64_t lsn, WT_PAGE_LOG_PUT_ARGS *args,
+      const WT_ITEM *buf)
+    {
+        auto acc = request<Pages, AccessMode::WRITE>(table_id);
+        acc.table.put(acc.conn, table_id, page_id, lsn, args, buf);
+        object_puts++;
     }
 
-    uint64_t
-    get_prev_full_page_lsn(Connection &conn, uint64_t table_id, uint64_t page_id, uint64_t lsn)
+    int
+    get_pages(uint64_t table_id, uint64_t page_id, WT_PAGE_LOG_GET_ARGS *args, uint32_t *flags,
+      WT_ITEM *results_array, uint32_t *results_count)
     {
-        StatementPtr stmt = conn.db_statement(Statement::GET_FULL_PAGE_LSN);
-        SQ_CHECK(sqlite3_bind_int64, stmt.get(), 1, static_cast<sqlite3_int64>(table_id));
-        SQ_CHECK(sqlite3_bind_int64, stmt.get(), 2, static_cast<sqlite3_int64>(page_id));
-        SQ_CHECK(sqlite3_bind_int64, stmt.get(), 3, static_cast<sqlite3_int64>(lsn));
-
-        int ret = SQ_CHECK(sqlite3_step, stmt.get());
-        if (ret == SQLITE_DONE) {
-            return 0; // No previous full page found
-        }
+        auto acc = request<Pages, AccessMode::READ>(table_id);
+        const int ret =
+          acc.table.get(acc.conn, table_id, page_id, args, flags, results_array, results_count);
+        acc.release();
```

Comment:
AFAIS that's the only place where we use release() for acces, could we avoid having it there and limit the access life time by its visibility scope? Can we remove the release() function in this case?


```suggestion
        {
            auto acc = request<Pages, AccessMode::READ>(table_id);
        const int ret =
             acc.table.get(acc.conn, table_id, page_id, args, flags, results_array, results_count);
        }
```

Reply:
Can we still remove the release() completely? Or at least make it public and use scope-locks everywhere?